ledger-tool/src/program.rs

```diff
@@ -131,18 +131,61 @@ fn load_blockstore(ledger_path: &Path, arg_matches: &ArgMatches<'_>) -> Arc<Bank
     bank
 }
 
-pub trait RunSubCommand {
-    fn run_subcommand(self) -> Self;
+pub trait ProgramSubCommand {
+    fn program_subcommand(self) -> Self;
 }
 
-impl RunSubCommand for App<'_, '_> {
-    fn run_subcommand(self) -> Self {
+impl ProgramSubCommand for App<'_, '_> {
+    fn program_subcommand(self) -> Self {
         self.subcommand(
+            SubCommand::with_name("program")
+        .about("Run to test, debug, and analyze on-chain programs.")
+        .subcommand(
+            SubCommand::with_name("cfg")
+                .about("generates Control Flow Graph of the program.")
+                .arg(
+                    Arg::with_name("PROGRAM")
```

Comment:
You can add this arg under the overall subcommand `program` to remove the duplication between `cfg`, `disassemble`, and `run`

Reply:
that doesn't work somehow. I did it, and received an error 
```
error: The argument 'max_genesis_archive_unpacked_size' wasn't found
```

Reply:
Here's how `program run --help` looks now

```
The command executes on-chain programs in a mocked environment.

The input data for a program execution have to be in JSON format
and the following fields are required
{
    "program_id": "DozgQiYtGbdyniV2T74xMdmjZJvYDzoRFFqw7UR5MwPK",
    "accounts": [
        {
            "key": "524HMdYYBy6TAn4dK5vCcjiTmT2sxV6Xoue5EXrz22Ca",
            "owner": "BPFLoaderUpgradeab1e11111111111111111111111",
            "is_signer": false,
            "is_writable": true,
            "lamports": 1000,
            "data": [0, 0, 0, 3]
        }
    ],
    "instruction_data": [31, 32, 23, 24]
}

USAGE:
    solana-ledger-tool program run [FLAGS] [OPTIONS] <PROGRAM>

FLAGS:
        --force-update-to-open    Allow commands that would otherwise not alter the blockstore to make necessary updates
                                  in order to open it
    -h, --help                    Prints help information
    -V, --version                 Prints version information
    -v, --verbose                 Show additional information where supported

OPTIONS:
        --incremental-snapshot-archive-path <DIR>       Use DIR for separate incremental snapshot location
    -i, --input <FILE / BYTES>
            Input for the program to run on, where FILE is a name of a JSON file with input data, or BYTES is the number
            of 0-valued bytes to allocate for program parameters [default: 0]
        --limit <COUNT>
            Limit the number of instructions to execute [default: 9223372036854775807]

    -l, --ledger <DIR>                                  Use DIR as ledger location [default: ledger]
        --max-genesis-archive-unpacked-size <NUMBER>
            maximum total uncompressed size of unpacked genesis archive [default: 10485760]

    -m, --memory <BYTES>                                Heap memory for the program to run on [default: 0]
    -e, --mode <VALUE>
            Mode of execution, where 'interpreter' runs the program in the virtual machine's interpreter, 'debugger' is
            the same as 'interpreter' but hosts a GDB interface, and 'jit' precompiles the program to native machine
            code before execting it in the virtual machine. [default: jit]  [possible values: interpreter, debugger,
            jit]
        --output <FORMAT>
            Return information in specified output format, currently only available for bigtable and program subcommands
            [possible values: json, json-compact]
        --port <PORT>
            Port to use for the connection with a remote debugger [default: 9001]

        --snapshot-archive-path <DIR>                   Use DIR for snapshot location
    -t, --trace <FILE>                                  Output instruction trace
        --wal-recovery-mode <MODE>
            Mode to recovery the ledger db write ahead log [possible values: tolerate_corrupted_tail_records,
            absolute_consistency, point_in_time, skip_any_corrupted_record]

ARGS:
    <PROGRAM>    Program file to use. This is either an ELF shared-object file to be executed, or an assembly file
                 to be assembled and executed.

```

Here's how it looks when I move <PROGRAM> argument to program subcommand

```
The command executes on-chain programs in a mocked environment.

The input data for a program execution have to be in JSON format
and the following fields are required
{
    "program_id": "DozgQiYtGbdyniV2T74xMdmjZJvYDzoRFFqw7UR5MwPK",
    "accounts": [
        {
            "key": "524HMdYYBy6TAn4dK5vCcjiTmT2sxV6Xoue5EXrz22Ca",
            "owner": "BPFLoaderUpgradeab1e11111111111111111111111",
            "is_signer": false,
            "is_writable": true,
            "lamports": 1000,
            "data": [0, 0, 0, 3]
        }
    ],
    "instruction_data": [31, 32, 23, 24]
}

USAGE:
    solana-ledger-tool program <PROGRAM> run [FLAGS] [OPTIONS]

FLAGS:
        --force-update-to-open    Allow commands that would otherwise not alter the blockstore to make necessary updates
                                  in order to open it
    -h, --help                    Prints help information
    -V, --version                 Prints version information
    -v, --verbose                 Show additional information where supported

OPTIONS:
        --incremental-snapshot-archive-path <DIR>       Use DIR for separate incremental snapshot location
    -i, --input <FILE / BYTES>
            Input for the program to run on, where FILE is a name of a JSON file with input data, or BYTES is the number
            of 0-valued bytes to allocate for program parameters [default: 0]
        --limit <COUNT>
            Limit the number of instructions to execute [default: 9223372036854775807]

    -l, --ledger <DIR>                                  Use DIR as ledger location [default: ledger]
        --max-genesis-archive-unpacked-size <NUMBER>
            maximum total uncompressed size of unpacked genesis archive [default: 10485760]

    -m, --memory <BYTES>                                Heap memory for the program to run on [default: 0]
    -e, --mode <VALUE>
            Mode of execution, where 'interpreter' runs the program in the virtual machine's interpreter, 'debugger' is
            the same as 'interpreter' but hosts a GDB interface, and 'jit' precompiles the program to native machine
            code before execting it in the virtual machine. [default: jit]  [possible values: interpreter, debugger,
            jit]
        --output <FORMAT>
            Return information in specified output format, currently only available for bigtable and program subcommands
            [possible values: json, json-compact]
        --port <PORT>
            Port to use for the connection with a remote debugger [default: 9001]

        --snapshot-archive-path <DIR>                   Use DIR for snapshot location
    -t, --trace <FILE>                                  Output instruction trace
        --wal-recovery-mode <MODE>
            Mode to recovery the ledger db write ahead log [possible values: tolerate_corrupted_tail_records,
            absolute_consistency, point_in_time, skip_any_corrupted_record]

```
and this output of `program --help`

```
Run to test, debug, and analyze on-chain programs.

USAGE:
    solana-ledger-tool program [FLAGS] [OPTIONS] <PROGRAM> [SUBCOMMAND]

FLAGS:
        --force-update-to-open    Allow commands that would otherwise not alter the blockstore to make necessary updates
                                  in order to open it
    -h, --help                    Prints help information
    -v, --verbose                 Show additional information where supported

OPTIONS:
        --incremental-snapshot-archive-path <DIR>    Use DIR for separate incremental snapshot location
    -l, --ledger <DIR>                               Use DIR as ledger location [default: ledger]
        --output <FORMAT>
            Return information in specified output format, currently only available for bigtable and program subcommands
            [possible values: json, json-compact]
        --snapshot-archive-path <DIR>                Use DIR for snapshot location
        --wal-recovery-mode <MODE>
            Mode to recovery the ledger db write ahead log [possible values: tolerate_corrupted_tail_records,
            absolute_consistency, point_in_time, skip_any_corrupted_record]

ARGS:
    <PROGRAM>    Program file to use. This is either an ELF shared-object file to be executed, or an assembly file
                 to be assembled and executed.

SUBCOMMANDS:
    cfg            generates Control Flow Graph of the program.
    disassemble    dumps disassembled code of the program.
    help           Prints this message or the help of the given subcommand(s)
    run            The command executes on-chain programs in a mocked environment.
                   
                   The input data for a program execution have to be in JSON format
                   and the following fields are required
                   {
                       "program_id": "DozgQiYtGbdyniV2T74xMdmjZJvYDzoRFFqw7UR5MwPK",
                       "accounts": [
                           {
                               "key": "524HMdYYBy6TAn4dK5vCcjiTmT2sxV6Xoue5EXrz22Ca",
                               "owner": "BPFLoaderUpgradeab1e11111111111111111111111",
                               "is_signer": false,
                               "is_writable": true,
                               "lamports": 1000,
                               "data": [0, 0, 0, 3]
                           }
                       ],
                       "instruction_data": [31, 32, 23, 24]
                   }

```

I prefer to see the description of `<PROGRAM>` argument when I request help for a sub-sub-command as is the case now.

Reply:
It looks like clap shows in subcommands inherited `--` (long or short) arguments, but not free standing arguments (ones without `-` or `--` fixed text option).

Reply:
In clap terminology it shows inherited options and flags, but not args.

Reply:
Done.